### PR TITLE
Adds skeleton and initial validation

### PR DESCRIPTION
**Pull request by stevenjack**
### Problem

It would be nice to simply validate our terraform stacks against the structure we define so we can easily run this in CI.

### Solution

* Add binary that validates that resources are split into correct files and that the files contain the correct resources.

### Notes

* Test will come next along with README update.

### CODE REVIEW

**Comment by stevenjack** [comment]
https://app.clubhouse.io/vidsy/story/6762

File: cmd/terraform/linter/main.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+)
+
+func main() {
+	tfDirectory := flag.String("tf-directory", "", "The directory that contains the terraform files to lint")
```
Comment (revett) [inline cmd/terraform/linter/main.go]: Multi-line

File: cmd/terraform/linter/main.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package main
+
+import (
+	"errors"
```
Comment (revett) [inline cmd/terraform/linter/main.go]: Can we use `pkg/errors`

Comment (stevenjack) [reply cmd/terraform/linter/main.go]: wasn't sure there was much point having stack traces for a cli but can do

File: cmd/terraform/linter/main.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		log.Fatal(err)
+	}
+
+	err = LintDirectory(*tfDirectory, files)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func isValidDirectory(path string) (bool, error) {
```
Comment (revett) [inline cmd/terraform/linter/main.go]: No need for `bool` as consuming code can use `err != nil` as a check for if the directory is valid.

File: cmd/terraform/linter/lint_directory.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/terraform/config"
+)
+
+// LintDirectory takes a set files an lints them based on the Vidsy
```
Comment (revett) [inline cmd/terraform/linter/lint_directory.go]: `an`

**Comment by revett** [comment]
Can a `README` be added that outlines the `rules` that the linter uses.

File: cmd/terraform/linter/lint_directory.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+// structure of stacks.
+func LintDirectory(directory string, files []os.FileInfo) error {
+	for _, file := range files {
+		if !validTFFile(file) {
```
Comment (revett) [inline cmd/terraform/linter/lint_directory.go]: Prefix function with `is...`

**Comment by stevenjack** [comment]
@revett See description, was going to add in next PR.

File: cmd/terraform/linter/lint_directory.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"errors"
```
Comment (revett) [inline cmd/terraform/linter/lint_directory.go]: Use `pkg/errors`

File: cmd/terraform/linter/lint_directory.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	if len(terraformConfig.ProviderConfigs) > 0 {
+		return fmt.Errorf(
+			"contains %d provider resource(s), these should be placed in 'providers.tf'",
+			len(terraformConfig.ProviderConfigs),
+		)
+	}
+
+	if len(terraformConfig.Resources) == 0 && len(terraformConfig.Modules) == 0 && len(terraformConfig.Locals) == 0 {
```
Comment (revett) [inline cmd/terraform/linter/lint_directory.go]: Multi-line

**Review by revett — CHANGES_REQUESTED** [review]
Added comments to code and the base PR as well

File: cmd/terraform/linter/lint_directory.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		if err != nil {
+			return fmt.Errorf("%s - %s", file.Name(), err)
+		}
+	}
+
+	return nil
+
+}
+
+func lintData(terraformConfig *config.Config) error {
```
Comment (pocockn) [inline cmd/terraform/linter/lint_directory.go]: Would the param be better called `conf` or something shorter? `terraformConfig.Terraform` seems quite verbose with no real gain.

Comment (stevenjack) [reply cmd/terraform/linter/lint_directory.go]: @pocockn like it 👍

**Review by pocockn — APPROVED** [review]
🍰 🔥

File: cmd/terraform/linter/main.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	tfDirectory := flag.String(
+		"tf-directory",
+		"",
+		"The directory that contains the terraform files to lint",
+	)
+	flag.Parse()
+
+	err := isValidDirectory(*tfDirectory)
+	if err != nil {
+		log.Fatal(err)
```
Comment (revett) [inline cmd/terraform/linter/main.go]: I think the following is needed to get the stack trace from the errors pkg:

```go
log.Fatalf("%+v", err)
```

**Review by revett — COMMENTED** [review]
1 small comment and then good to go

**Review by revett — APPROVED** [review]
🍰

**Pull request merged**